### PR TITLE
fix(#241): updateDai does not create DOI SDI DAI as expected

### PR DESCRIPTION
**Pull request by massifben**

### CODE REVIEW

**Review by AliouDIAITE — APPROVED** [review]

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/LNAdapterTest.java [hunk 1]
```diff
@@ -505,33 +504,81 @@ void updateDAI_should_not_update_DAI_Val_when_DTT_Fc_not_defined() {
...
+        // Then
+        System.out.println(MarshallerWrapper.marshall(scd));
+        SDIAdapter.DAIAdapter daiAdapter = lnAdapter
+                .getDOIAdapterByName("Do")
+                .getStructuredDataAdapterByName("sdo1")
+                .getStructuredDataAdapterByName("d")
+                .getStructuredDataAdapterByName("antRef")
+                .getStructuredDataAdapterByName("bda1")
+                .getStructuredDataAdapterByName("bda2")
+                .getDataAdapterByName("bda3");
```
Comment (samirromdhani) [inline sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/LNAdapterTest.java:543]: Only object of type `TDAI` can have value and `TSDI` can only have `TSDI` or `TDAI` .
The expected result should be: 
```java
SDIAdapter.DAIAdapter daiAdapter = lnAdapter
               .getDOIAdapterByName("Do") //name of TDOI
                .getStructuredDataAdapterByName("sdo1") //name of TSDI
                .getStructuredDataAdapterByName("d") //name of TSDI
                .getStructuredDataAdapterByName("bda1") //name of TSDI
                .getStructuredDataAdapterByName("bda2") //name of TSDI
                .getStructuredDataAdapterByName("bda3") //name of TSDI
                .getDataAdapterByName("antRef");//name of TDAI
assertThat(daiAdapter.getCurrentElem().getName()).isEqualTo("antRef");
assertThat(daiAdapter.getCurrentElem().getVal().get(0).getValue()).isEqualTo("newValue");
```
As mentioned in https://github.com/com-pas/compas-sct/issues/241
```xml
<DOI name="doName">
    <SDI name="sdoName1">
        <SDI name="sdoName2">
            <SDI name="bdaName1">
                <SDI name="bdaName2">
                    <SDI name="bdaName3">
                        <DAI name="daName" valImport="true">
                            <Val>newValue</Val>
                        </DAI>
                    </SDI>
                </SDI>
            </SDI>
        </SDI>
</DOI>
```

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java [hunk 2]
```diff
@@ -685,74 +685,88 @@ protected Optional<IDataAdapter> findMatch(DoTypeName doTypeName, DaTypeName daT
...
 
-        if (!rDtt.isUpdatable())
-            return;
+    private IDataParentAdapter createDoiAndSdiForDoAndSdoIfNotExists(DAITracker daiTracker) {
+        DoTypeName doTypeName = daiTracker.getDoTypeName();
+        IDataParentAdapter parent = daiTracker.getDoiOrSdoiAdapter();
+        if (parent == null) { // No DOI found, same as (matchResult == MatchResult.FAILED), same as (getIndexDoType() == DOI_NOT_FOUND)
+            parent = addDOI(doTypeName.getName());
+        }
+        int existingSdiForSdo = daiTracker.getIndexDoType() < 0 ? 0 : daiTracker.getIndexDoType() + 1;
```
Comment (samirromdhani) [inline sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java:740]:
```suggestion
        int instantiatedSDICount = daiTracker.getIndexDoType() < 0 ? 0 : daiTracker.getIndexDoType() + 1;
```

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java [hunk 3]
```diff
@@ -685,74 +685,88 @@ protected Optional<IDataAdapter> findMatch(DoTypeName doTypeName, DaTypeName daT
...
-                    daiOrBdaiAdapter = doiOrSdoiAdapter.addSDOI(daTypeName.getName());
-                } else if (i == daSz - 1) {
-                    daiAdapter = daiOrBdaiAdapter.addDAI(bdaName, rDtt.isUpdatable());
-                } else {
-                    daiOrBdaiAdapter = daiOrBdaiAdapter.addSDOI(bdaName);
-                }
-            }
-            if (daiAdapter == null) {
-                daiAdapter = doiOrSdoiAdapter.addDAI(daTypeName.getName(), rDtt.isUpdatable());
+            int existingSdiForBda = daiTracker.getIndexDaType() < 0 ? 0 : daiTracker.getIndexDaType() + 1;
```
Comment (samirromdhani) [inline sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java:760]:
```suggestion
            int instantiatedSDICount = daiTracker.getIndexDaType() < 0 ? 0 : daiTracker.getIndexDaType() + 1;
```

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java [hunk 4]
```diff
@@ -685,74 +685,88 @@ protected Optional<IDataAdapter> findMatch(DoTypeName doTypeName, DaTypeName daT
...
+     * It does not even check if rDtt exists in DataTypeTemplate section.
+     * That means that it will create the missing DOI/SDI/DAI, even if it is not consistent with DataTypeTemplate section.
+     * It is the caller responsibility to ensure the consistency between the given rDtt and the lnType of this LN/LN0 (which points to DataTypeTemplate section).
+     * See 9.3.5 "LN0 and other Logical Nodes" of IEC 61850-6.
+     * @param doTypeName DO to match with DOI and SDO (if any) to match SDI
+     * @param daTypeName DA to match DAI when no BDA, otherwise DA and BDA to match SDI and DAI.
+     * @param setValImportOnCreate when this method creates the DAI, it will set DAI.valImport attribute with this parameter
+     * @return adapter for existing DAI or created DAI.
+     */
+    private AbstractDAIAdapter<?> createDaiAdapterIfNotExists(DoTypeName doTypeName, DaTypeName daTypeName, boolean setValImportOnCreate) {
```
Comment (samirromdhani) [inline sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java:727]: Make it public to facilitate the testing of various scenarios.

**Review by samirromdhani — APPROVED** [review]

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java (hunk 2 above)
Comment (massifben) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java:740]: OK

**Review by SaintierFr — APPROVED** [review]

File: sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java (hunk 4 above)
Comment (gleizesDor) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java:727]: I'm not really agree, 
if this method is dedicated to be private then it should be private.  
Tests can be readable by other ways

Comment (samirromdhani) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java:727]: @massifben I think you plan to update this PR by this branch **241-updatedai-val-does-not-produce-subelements-sdi-as-expected_2** where no challenging class exit like DAITracker.

Comment (massifben) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java]: I rebased `241-updatedai-val-does-not-produce-subelements-sdi-as-expected` on `241-updatedai-val-does-not-produce-subelements-sdi-as-expected_2` and forced push. So the PR is up to date.

Comment (massifben) [reply sct-commons/src/main/java/org/lfenergy/compas/sct/commons/scl/ied/AbstractLNAdapter.java]: So we agree to keep the method private. I will test it through `updateDai`

> I'm not really agree, if this method is dedicated to be private then it should be private. Tests can be readable by other ways

So I keep it private (I need to adapt the the tests before closing this thread)

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/LNAdapterTest.java [hunk 5]
```diff
@@ -572,15 +599,165 @@ void updateDAI_should_not_update_DAI_values_when_not_updatable() {
...
+
+    @Test
+    void updateDAI_with_SDO_and_BDA_when_only_DAI_is_missing_should_create_DAI() {
+        // Given
+        SCL scd = SclTestMarshaller.getSCLFromFile("/ied-test-schema-conf/ied_unit_test.xml");
+        AbstractLNAdapter<?> lnAdapter = findLn0(new SclRootAdapter(scd), "IED_NAME", "LD_INS1");
+        AbstractDAIAdapter<?> daiToRemove = findDai(lnAdapter, DATA_TYPE_REF);
+        ((IDataParentAdapter) daiToRemove.getParentAdapter()).getSDIOrDAI().clear();
+
+        System.out.println("BEFORE => " + MarshallerWrapper.marshall(scd));
```
Comment (samirromdhani) [inline sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/LNAdapterTest.java:730]: Before and after prints are useless.
You can use something similar to this instead. 
```java
// Given
...
var before = lnAdapter.findDoiAdapterByName(DO_TYPE_NAME.getName()).get()
        .getStructuredDataAdapterByName("sdo1")
        .getStructuredDataAdapterByName("d")
        .getStructuredDataAdapterByName("antRef")
        .getStructuredDataAdapterByName("bda1")
        .getStructuredDataAdapterByName("bda2");
assertThat(before.getSDIOrDAI().size()).isEqualTo(0);
// When
lnAdapter.updateDAI(RDDT_DO_SDO1_D_ANTREF_BDA1_BDA2_BDA3);
// Then
assertThat(before.getSDIOrDAI().size()).isEqualTo(1);
var expected = before.getDataAdapterByName("bda3");
assertThat(expected.getCurrentElem().getVal().get(0).getValue()).isEqualTo(NEW_VALUE);
...
```

**Review by samirromdhani — APPROVED** [review]

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/LNAdapterTest.java (hunk 5 above)
Comment (massifben) [reply sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/LNAdapterTest.java]: OK. I removed the print BEFORE and AFTER and added an assertion at the end of the  `//Given` part

File: sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/LNAdapterTest.java [hunk 6]
```diff
@@ -572,15 +598,152 @@ void updateDAI_should_not_update_DAI_values_when_not_updatable() {
...
+        SCL scd = SclTestMarshaller.getSCLFromFile("/ied-test-schema-conf/ied_unit_test.xml");
+        AbstractLNAdapter<?> lnAdapter = findLn0(new SclRootAdapter(scd), "IED_NAME", "LD_INS1");
+        lnAdapter.getDOIAdapterByName(DO_TYPE_NAME.getName()).getCurrentElem().getSDIOrDAI().clear();
+        assertThat(findDoiOrSdi(lnAdapter, DO_TYPE_NAME.getName()).getSDIOrDAI()).isEmpty();
+
+        // When
+        lnAdapter.updateDAI(RDDT_DO_SDO1_D_ANTREF_BDA1_BDA2_BDA3);
+
+        // Then
+        System.out.println("AFTER => " + MarshallerWrapper.marshall(scd));
```
Comment (samirromdhani) [inline sct-commons/src/test/java/org/lfenergy/compas/sct/commons/scl/ied/LNAdapterTest.java:686]: forgotten print here.

**Pull request merged**